Clarify URL requirements (#94)
APIs-guru/graphql-over-http: README.md

# Overview

Serving GraphQL over HTTP provides the ability to use the full advantages of GraphQL with the rich feature set of HTTP.
Carrying GraphQL in HTTP does not mean that GraphQL overrides existing semantics of HTTP,
but rather that the semantics of GraphQL over HTTP map naturally to HTTP semantics.

GraphQL naturally follows the HTTP request/response message model,
providing a GraphQL request in an HTTP request and GraphQL response in an HTTP response.

# URL

A GraphQL over HTTP compliant server MUST designate at least one URL that handles GraphQL requests.

> It is allowed for the same GraphQL schema to be available on multiple URLs.

Reviewer: I wonder if this line is needed (I doubt that many people have a need to run the same schema more than once on a server).  It might help to be more generic.  Something that tells people that each URL that handles GraphQL requests MUST be an independent GraphQL endpoint (even one with the same schema), as long as they are each complete.

How about:

```
Multiple URLs MAY exist on a server to handle GraphQL requests.  These URLs MUST be independent and complete.
```

Author: I think this is valuable to avoid ambiguity in scenarios where a server migrates its GraphQL endpoint to a different URL, offering both in a transitional period.

Author: Do you think the sentence above covers the completeness aspect?

> All GraphQL operations that are available in the schema MUST be available on the URLs dedicated to it.

Author: Added the first part of your suggestion.

I'd rather have the spec be a bit more wordy and possibly redundant than to be ambiguous.

All GraphQL operations that are available in the schema MUST be available on each designated URL.

> That means the client can send all their GraphQL requests to a single endpoint, regardless

Author: I think this is the reasoning for serving the entire schema on a single URL, it is valuable to mention that in the spec.

Reviewer 2: I'd rather first stress that a dedicated URL has to support everything in "its" schema, so clients can send all requests based on that schema to the same URL.

And then point out that there may be multiple URLs, each serving a potentially (but not necessarily) different schema.

Author: > I'd rather first stress that a dedicated URL has to support everything in "its" schema, so clients can send all requests based on that schema to the same URL.

I tweaked the wording to stress that bit.

> And then point out that there may be multiple URLs, each serving a potentially (but not necessarily) different schema.

This was also suggested by @ghmcadams, i added a note saying that multiple URLs may exist.

of the contained Operations, Fragments or Variables.

> Other protocols may also use those URLs.

Reviewer: I'm not sure `protocols` is the right word here.

Author: I expanded that bit.

# Overview

Serving GraphQL over HTTP provides the ability to use the full advantages of GraphQL with the rich feature set of HTTP.
Carrying GraphQL in HTTP does not mean that GraphQL overrides existing semantics of HTTP,
but rather that the semantics of GraphQL over HTTP map naturally to HTTP semantics.

GraphQL naturally follows the HTTP request/response message model,
providing a GraphQL request in an HTTP request and GraphQL response in an HTTP response.

# URL

A GraphQL over HTTP compliant server MUST designate at least one URL that handles GraphQL requests.

> All GraphQL operations that are available in a schema MUST be available on all URLs that are designated to it.

Author: TODO for myself: Work in the relation between schema and introspection.

> All GraphQL operations that are available in a schema MUST be available on all URLs that are designated to it.

Author: ```suggestion
All Queries and Mutations that are available in a schema MUST be available on all URLs that are designated to it.
```

Reviewer 3: What about subscriptions?

Author: Subscriptions often have some kind of redirection going on where multiple servers are involved. We might specify that in the future, but don't have a clear answer now.

That means the client can send all their GraphQL requests to a single endpoint, regardless
of the contained Operations, Fragments or Variables.

Multiple URLs MAY exist on a server to handle GraphQL requests, potentially serving different

> but independent schemas. The same GraphQL schema MAY be available on multiple URLs on the server.

Author: ```suggestion
but standalone schemas. The same GraphQL schema MAY be available on multiple URLs on the server.
```

Other ideas:
- Self-sufficient
- Complete

Maybe we don't need this section at all, but would rather specify cleanly what a single URL does.

> but independent schemas. It is allowed for the same GraphQL schema to be available on multiple URLs.

Reviewer: ```suggestion
but independent schemas. The same GraphQL schema MAY be available on multiple URLs on the server.
```

> Other protocols or services may also use those URLs, as long as they don't conflict with the
> server's responsibility to handle GraphQL requests.

Reviewer: I believe `protocol` is not the right word here.  GraphQL is not a protocol.  HTTP is a protocol.  How about this:

```suggestion
Those URLs MAY also handle other types of requests, as long as they don't conflict with the
server's responsibility to handle GraphQL requests.
```

Author: Well, it means the URL may also be used to handle FTP or other protocols. Maybe this would be even for generic:

```suggestion
Those URLs MAY also be used for other purposes, as long as they don't conflict with the
server's responsibility to handle GraphQL requests.
```

Reviewer: This wording is much better.
